core/trino-main/src/main/java/io/trino/server/security/oauth2/NimbusAirliftHttpClient.java

```diff
@@ -0,0 +1,149 @@
+        }
+        catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public <T> T execute(com.nimbusds.oauth2.sdk.Request nimbusRequest, Parser<T> parser)
+    {
+        HTTPRequest httpRequest = nimbusRequest.toHTTPRequest();
+        HTTPRequest.Method method = httpRequest.getMethod();
+
+        Request.Builder request = new Request.Builder()
+                .setMethod(method.name())
+                .setFollowRedirects(httpRequest.getFollowRedirects());
+
+        UriBuilder url = UriBuilder.fromUri(httpRequest.getURI());
+        if (method.equals(GET) || method.equals(DELETE)) {
+            httpRequest.getQueryParameters().forEach((key, value) -> url.queryParam(key, value.toArray()));
+        }
+
+        url.fragment(httpRequest.getFragment());
+
+        request.setUri(URI.create(url.toString()));
+
+        ImmutableMultimap.Builder<String, String> headers = ImmutableMultimap.builder();
+        httpRequest.getHeaderMap().forEach(headers::putAll);
+        request.addHeaders(headers.build());
+
+        if (method.equals(POST) || method.equals(PUT)) {
+            String query = httpRequest.getQuery();
+            if (query != null) {
+                request.setBodyGenerator(createStaticBodyGenerator(httpRequest.getQuery(), UTF_8));
+            }
+        }
+        return httpClient.execute(request.build(), new NimbusResponseHandler<>(parser));
+    }
+
+    public static class NimbusResponseHandler<T>
+            implements ResponseHandler<T, RuntimeException>
+    {
+        private final StringResponseHandler handler = createStringResponseHandler();
+        private final Parser<T> parser;
+
+        public NimbusResponseHandler(Parser<T> parser)
+        {
+            this.parser = requireNonNull(parser, "parser is null");
+        }
+
+        @Override
+        public T handleException(Request request, Exception exception)
+        {
+            if (exception instanceof ConnectException) {
+                throw new RuntimeException("Server refused connection: " + urlFor(request), exception);
+            }
+            if (exception instanceof IOException) {
+                throw new RuntimeException("Failed communicating with server: " + urlFor(request), exception);
+            }
+            throw new RuntimeException("Unexpected error occurred: " + urlFor(request), exception);
+        }
+
+        @Override
+        public T handle(Request request, Response response)
+        {
+            StringResponseHandler.StringResponse stringResponse = handler.handle(request, response);
+            HTTPResponse nimbusResponse = new HTTPResponse(response.getStatusCode());
+            response.getHeaders().asMap().forEach((name, values) -> nimbusResponse.setHeader(name.toString(), values.toArray(new String[0])));
+            nimbusResponse.setContent(stringResponse.getBody());
+            try {
+                return parser.parse(nimbusResponse);
```

Comment:
It's because `Parser` interface represents parsing methods convention of Nimbus library, for example for `UserInfoResponse`:
```java
public static UserInfoResponse parse(final HTTPResponse httpResponse) {}
```
in which case we can simply pass `UserInfoResponse::parse` to `NimbusHttpClient#execute` method.